File: cpp_generated/agent-portability-bun.test.cpp
#include "/home/runner/work/elizaos-cpp/elizaos-cpp/classified/packages/agentserver/src/__tests__/integration/agent-portability-bun.test.h"

void Main(void)
{
    describe(std::string("Agent Portability Integration"), [=]() mutable
    {
        shared testAgentId = as<std::shared_ptr<UUID>>(std::string("test-agent-123"));
        shared createMockServer = [=](auto agentExists = true) mutable
        {
            shared mockDb = object{
                object::pair{std::string("select"), [=]() mutable
                {
                    return mockDb;
                }
                }, 
                object::pair{std::string("from"), [=]() mutable
                {
                    return mockDb;
                }
                }, 
                object::pair{std::string("where"), [=]() mutable
                {
                    return Promise->resolve((agentExists) ? array<object>{ object{
                        object::pair{std::string("id"), testAgentId}, 
                        object::pair{std::string("name"), std::string("Test Agent")}
                    } } : array<any>());
                }
                }, 
                object::pair{std::string("insert"), [=]() mutable
                {
                    return mockDb;
                }
                }, 
                object::pair{std::string("values"), [=]() mutable
                {
                    return Promise->resolve();
                }
                }, 
                object::pair{std::string("delete"), [=]() mutable
                {
                    return mockDb;
                }
                }, 
                object::pair{std::string("transaction"), [=](auto fn) mutable
                {
                    shared tx = object{
                        object::pair{std::string("select"), [=]() mutable
                        {
                            return tx;
                        }
                        }, 
                        object::pair{std::string("from"), [=]() mutable
                        {
                            return tx;
                        }
                        }, 
                        object::pair{std::string("where"), [=]() mutable
                        {
                            return Promise->resolve(array<any>());
                        }
                        }, 
                        object::pair{std::string("insert"), [=]() mutable
                        {
                            return tx;
                        }
                        }, 
                        object::pair{std::string("values"), [=]() mutable
                        {
                            return Promise->resolve();
                        }
                        }, 
                        object::pair{std::string("delete"), [=]() mutable
                        {
                            return tx;
                        }
                        }
                    };
                    return std::async([=]() { fn(tx); });
                }
                }
            };
            return as<any>(object{
                object::pair{std::string("db"), mockDb}
            });
        };
        describe(std::string("Export Service"), [=]() mutable
        {
            it(std::string("should create a valid export ZIP structure"), [=]() mutable
            {
                auto mockServer = createMockServer(true);
                auto mockRuntime = as<std::shared_ptr<IAgentRuntime>>(object{});
                shared queryCount = 0;
                mockServer["db"]["where"] = [=]() mutable
                {
                    queryCount++;
                    auto responses = array<array<object>>{ array<object>{ object{
                        object::pair{std::string("id"), testAgentId}, 
                        object::pair{std::string("name"), std::string("Test Agent")}, 
                        object::pair{std::string("enabled"), true}
                    } }, array<object>{ object{
                        object::pair{std::string("id"), std::string("entity-1")}, 
                        object::pair{std::string("agentId"), testAgentId}, 
                        object::pair{std::string("names"), array<string>{ std::string("Entity") }}
                    } }, array<any>(), array<any>(), array<any>(), array<object>{ object{
                        object::pair{std::string("id"), std::string("mem-1")}, 
                        object::pair{std::string("agentId"), testAgentId}, 
                        object::pair{std::string("content"), object{
                            object::pair{std::string("text"), std::string("Test")}
                        }}, 
                        object::pair{std::string("embedding"), array<double>{ 0.1, 0.2 }}
                    } }, array<any>(), array<any>(), array<any>() };
                    return Promise->resolve(OR((const_(responses)[queryCount - 1]), (array<any>())));
                };
                auto exportService = std::make_shared<AgentExportService>(testAgentId, mockRuntime, mockServer);
                try
                {
                    shared zipStream = std::async([=]() { exportService->exportToZip(); });
                    shared chunks = array<std::shared_ptr<Buffer>>();
                    std::async([=]() { std::make_shared<Promise<void>>([=](auto resolve, auto reject) mutable
                    {
                        zipStream->on(std::string("data"), [=](auto chunk) mutable
                        {
                            return chunks->push(chunk);
                        }
                        );
                        zipStream->on(std::string("end"), [=]() mutable
                        {
                            return resolve();
                        }
                        );
                        zipStream->on(std::string("error"), reject);
                    }
                    ); });
                    auto zipBuffer = Buffer::concat(chunks);
                    auto zip = std::make_shared<AdmZip>(zipBuffer);
                    auto entries = zip->getEntries();
                    auto entryNames = entries->map([=](auto e) mutable
                    {
                        return e["entryName"];
                    }
                    );
                    expect(entryNames)->toContain(std::string("manifest.json"));
                    expect(entryNames)->toContain(std::string("database/agent.json"));
                    expect(entryNames)->toContain(std::string("database/entities.json"));
                    expect(entryNames)->toContain(std::string("database/memories.json"));
                    auto manifestEntry = zip->getEntry(std::string("manifest.json"));
                    auto manifest = JSON->parse(zip->readAsText(manifestEntry));
                    expect(manifest["agentId"])->toBe(testAgentId);
                    expect(manifest["tables"]["length"])->toBe(9);
                    std::async([=]() { exportService->cleanup(); });
                }
                catch (const any& error)
                {
                    std::async([=]() { exportService->cleanup(); });
                    throw any(error);
                }
            }
            );
            it(std::string("should handle non-existent agent gracefully"), [=]() mutable
            {
                auto mockServer = createMockServer(false);
                auto mockRuntime = as<std::shared_ptr<IAgentRuntime>>(object{});
                auto exportService = std::make_shared<AgentExportService>(testAgentId, mockRuntime, mockServer);
                std::async([=]() { expect(exportService->exportToZip())->rejects->toThrow(std::string("Agent ") + testAgentId + std::string(" not found")); });
                std::async([=]() { exportService->cleanup(); });
            }
            );
        }
        );
        describe(std::string("Import Service"), [=]() mutable
        {
            it(std::string("should validate ZIP structure before import"), [=]() mutable
            {
                auto mockServer = createMockServer();
                auto invalidZip = std::make_shared<AdmZip>();
                invalidZip->addFile(std::string("test.txt"), Buffer::from(std::string("test")));
                auto importService = std::make_shared<AgentImportService>(testAgentId, mockServer);
                std::async([=]() { expect(importService->importFromZip(invalidZip->toBuffer()))->rejects->toThrow(std::string("Invalid export: manifest.json not found")); });
                std::async([=]() { importService->cleanup(); });
            }
            );
            it(std::string("should validate manifest version"), [=]() mutable
            {
                auto mockServer = createMockServer();
                auto zip = std::make_shared<AdmZip>();
                auto manifest = object{
                    object::pair{std::string("version"), std::string("99.0.0")}, 
                    object::pair{std::string("agentId"), std::string("source-agent")}, 
                    object::pair{std::string("agentName"), std::string("Test")}, 
                    object::pair{std::string("tables"), array<string>{ std::string("agent") }}, 
                    object::pair{std::string("fileCount"), 0}
                };
                zip->addFile(std::string("manifest.json"), Buffer::from(JSON->stringify(manifest)));
                zip->addFile(std::string("database/agent.json"), Buffer::from(std::string("[]")));
                auto importService = std::make_shared<AgentImportService>(testAgentId, mockServer);
                std::async([=]() { expect(importService->importFromZip(zip->toBuffer()))->rejects->toThrow(std::string("Unsupported export version: 99.0.0")); });
                std::async([=]() { importService->cleanup(); });
            }
            );
            it(std::string("should successfully import valid ZIP"), [=]() mutable
            {
                auto mockServer = createMockServer();
                auto zip = std::make_shared<AdmZip>();
                auto manifest = object{
                    object::pair{std::string("version"), std::string("1.0.0")}, 
                    object::pair{std::string("exportedAt"), ((std::make_shared<Date>()))->toISOString()}, 
                    object::pair{std::string("agentId"), std::string("source-agent")}, 
                    object::pair{std::string("agentName"), std::string("Test Agent")}, 
                    object::pair{std::string("tables"), array<string>{ std::string("agent"), std::string("entities"), std::string("worlds"), std::string("rooms"), std::string("participants"), std::string("memories"), std::string("relationships"), std::string("tasks"), std::string("server_agents") }}, 
                    object::pair{std::string("fileCount"), 0}, 
                    object::pair{std::string("elizaVersion"), std::string("1.0.0")}
                };
                zip->addFile(std::string("manifest.json"), Buffer::from(JSON->stringify(manifest)));
                auto agentData = array<object>{ object{
                    object::pair{std::string("id"), std::string("source-agent")}, 
                    object::pair{std::string("name"), std::string("Test Agent")}, 
                    object::pair{std::string("enabled"), true}, 
                    object::pair{std::string("createdAt"), ((std::make_shared<Date>()))->toISOString()}, 
                    object::pair{std::string("updatedAt"), ((std::make_shared<Date>()))->toISOString()}
                } };
                zip->addFile(std::string("database/agent.json"), Buffer::from(JSON->stringify(agentData)));
                auto emptyTables = array<string>{ std::string("entities"), std::string("worlds"), std::string("rooms"), std::string("participants"), std::string("memories"), std::string("relationships"), std::string("tasks"), std::string("server_agents") };
                for (auto& table : emptyTables)
                {
                    zip->addFile(std::string("database/") + table + std::string(".json"), Buffer::from(std::string("[]")));
                }
                auto importService = std::make_shared<AgentImportService>(testAgentId, mockServer);
                auto result = std::async([=]() { importService->importFromZip(zip->toBuffer(), object{
                    object::pair{std::string("overwrite"), true}, 
                    object::pair{std::string("validateManifest"), true}
                }); });
                expect(result->success)->toBe(true);
                expect(result->agentId)->toBe(testAgentId);
                expect(result->agentName)->toBe(std::string("Test Agent"));
                expect(result->tablesImported->get_length())->toBe(9);
                std::async([=]() { importService->cleanup(); });
            }
            );
        }
        );
        describe(std::string("Export/Import Cycle"), [=]() mutable
        {
            it(std::string("should maintain data integrity through export/import cycle"), [=]() mutable
            {
                auto exportZip = std::make_shared<AdmZip>();
                auto manifest = object{
                    object::pair{std::string("version"), std::string("1.0.0")}, 
                    object::pair{std::string("exportedAt"), ((std::make_shared<Date>()))->toISOString()}, 
                    object::pair{std::string("agentId"), std::string("cycle-test-agent")}, 
                    object::pair{std::string("agentName"), std::string("Cycle Test Agent")}, 
                    object::pair{std::string("tables"), array<string>{ std::string("agent"), std::string("entities"), std::string("worlds"), std::string("rooms"), std::string("participants"), std::string("memories"), std::string("relationships"), std::string("tasks"), std::string("server_agents") }}, 
                    object::pair{std::string("fileCount"), 1}, 
                    object::pair{std::string("elizaVersion"), std::string("1.0.0")}
                };
                exportZip->addFile(std::string("manifest.json"), Buffer::from(JSON->stringify(manifest, nullptr, 2)));
                auto agentData = array<object>{ object{
                    object::pair{std::string("id"), std::string("cycle-test-agent")}, 
                    object::pair{std::string("name"), std::string("Cycle Test Agent")}, 
                    object::pair{std::string("enabled"), true}, 
                    object::pair{std::string("bio"), array<string>{ std::string("Test bio line 1"), std::string("Test bio line 2") }}, 
                    object::pair{std::string("settings"), object{
                        object::pair{std::string("theme"), std::string("dark")}, 
                        object::pair{std::string("language"), std::string("en")}
                    }}, 
                    object::pair{std::string("metadata"), object{
                        object::pair{std::string("version"), 1}, 
                        object::pair{std::string("tags"), array<string>{ std::string("test"), std::string("cycle") }}
                    }}, 
                    object::pair{std::string("createdAt"), ((std::make_shared<Date>()))->toISOString()}, 
                    object::pair{std::string("updatedAt"), ((std::make_shared<Date>()))->toISOString()}
                } };
                exportZip->addFile(std::string("database/agent.json"), Buffer::from(JSON->stringify(agentData, nullptr, 2)));
                auto memories = array<object>{ object{
                    object::pair{std::string("id"), std::string("memory-1")}, 
                    object::pair{std::string("agentId"), std::string("cycle-test-agent")}, 
                    object::pair{std::string("content"), object{
                        object::pair{std::string("text"), std::string("Test memory with embedding")}
                    }}, 
                    object::pair{std::string("embedding"), ((array(768)))->fill(0.1)}, 
                    object::pair{std::string("metadata"), object{
                        object::pair{std::string("type"), std::string("test")}
                    }}, 
                    object::pair{std::string("createdAt"), Date->now()}
                } };
                exportZip->addFile(std::string("database/memories.json"), Buffer::from(JSON->stringify(memories, nullptr, 2)));
                auto emptyTables = array<string>{ std::string("entities"), std::string("worlds"), std::string("rooms"), std::string("participants"), std::string("relationships"), std::string("tasks"), std::string("server_agents") };
                for (auto& table : emptyTables)
                {
                    exportZip->addFile(std::string("database/") + table + std::string(".json"), Buffer::from(std::string("[]")));
                }
                exportZip->addFile(std::string("uploads/test.txt"), Buffer::from(std::string("test file content")));
                auto zipBuffer = exportZip->toBuffer();
                auto verifyZip = std::make_shared<AdmZip>(zipBuffer);
                auto entries = verifyZip->getEntries();
                expect(entries->length)->toBe(11);
                auto manifestEntry = verifyZip->getEntry(std::string("manifest.json"));
                auto readManifest = JSON->parse(verifyZip->readAsText(manifestEntry));
                expect(readManifest["agentId"])->toBe(std::string("cycle-test-agent"));
                auto agentEntry = verifyZip->getEntry(std::string("database/agent.json"));
                auto readAgent = JSON->parse(verifyZip->readAsText(agentEntry));
                expect(const_(readAgent)[0]["bio"])->toEqual(array<string>{ std::string("Test bio line 1"), std::string("Test bio line 2") });
                expect(const_(readAgent)[0]["settings"]["theme"])->toBe(std::string("dark"));
                auto memoriesEntry = verifyZip->getEntry(std::string("database/memories.json"));
                auto readMemories = JSON->parse(verifyZip->readAsText(memoriesEntry));
                expect(const_(readMemories)[0]["embedding"]["length"])->toBe(768);
                expect(const_(readMemories)[0]["content"]["text"])->toBe(std::string("Test memory with embedding"));
                auto fileEntry = verifyZip->getEntry(std::string("uploads/test.txt"));
                expect(fileEntry->getData()->toString())->toBe(std::string("test file content"));
            }
            );
        }
        );
    }
    );
}

MAIN

File: classified/packages/agentserver/src/__tests__/integration/agent-portability-bun.test.ts
import { describe, it, expect } from 'bun:test';
import AdmZip from 'adm-zip';
import { AgentExportService } from '../../services/AgentExportService';
import { AgentImportService } from '../../services/AgentImportService';
import type { IAgentRuntime, UUID } from '@elizaos/core';
import type { AgentServer } from '../../server';

describe('Agent Portability Integration', () => {
  const testAgentId = 'test-agent-123' as UUID;

  // Mock server instance
  const createMockServer = (agentExists: boolean = true) => {
    const mockDb = {
      select: () => mockDb,
      from: () => mockDb,
      where: () => Promise.resolve(agentExists ? [{ id: testAgentId, name: 'Test Agent' }] : []),
      insert: () => mockDb,
      values: () => Promise.resolve(),
      delete: () => mockDb,
      transaction: async (fn: Function) => {
        const tx = {
          select: () => tx,
          from: () => tx,
          where: () => Promise.resolve([]),
          insert: () => tx,
          values: () => Promise.resolve(),
          delete: () => tx,
        };
        return await fn(tx);
      },
    };

    return {
      db: mockDb,
    } as any;
  };

  describe('Export Service', () => {
    it('should create a valid export ZIP structure', async () => {
      const mockServer = createMockServer(true);
      const mockRuntime = {} as IAgentRuntime;

      // Mock the database to return some data
      let queryCount = 0;
      mockServer.db.where = () => {
        queryCount++;
        const responses = [
          [{ id: testAgentId, name: 'Test Agent', enabled: true }], // agent
          [{ id: 'entity-1', agentId: testAgentId, names: ['Entity'] }], // entities
          [], // worlds
          [], // rooms
          [], // participants
          [{ id: 'mem-1', agentId: testAgentId, content: { text: 'Test' }, embedding: [0.1, 0.2] }], // memories
          [], // relationships
          [], // tasks
          [], // server_agents
        ];
        return Promise.resolve(responses[queryCount - 1] || []);
      };

      const exportService = new AgentExportService(testAgentId, mockRuntime, mockServer);

      try {
        const zipStream = await exportService.exportToZip();

        // Collect stream into buffer
        const chunks: Buffer[] = [];
        await new Promise<void>((resolve, reject) => {
          zipStream.on('data', (chunk) => chunks.push(chunk));
          zipStream.on('end', () => resolve());
          zipStream.on('error', reject);
        });

        const zipBuffer = Buffer.concat(chunks);

        // Verify ZIP structure
        const zip = new AdmZip(zipBuffer);
        const entries = zip.getEntries();
        const entryNames = entries.map((e) => e.entryName);

        expect(entryNames).toContain('manifest.json');
        expect(entryNames).toContain('database/agent.json');
        expect(entryNames).toContain('database/entities.json');
        expect(entryNames).toContain('database/memories.json');

        // Verify manifest
        const manifestEntry = zip.getEntry('manifest.json');
        const manifest = JSON.parse(zip.readAsText(manifestEntry!));
        expect(manifest.agentId).toBe(testAgentId);
        expect(manifest.tables.length).toBe(9);

        await exportService.cleanup();
      } catch (error) {
        await exportService.cleanup();
        throw error;
      }
    });

    it('should handle non-existent agent gracefully', async () => {
      const mockServer = createMockServer(false);
      const mockRuntime = {} as IAgentRuntime;

      const exportService = new AgentExportService(testAgentId, mockRuntime, mockServer);

      await expect(exportService.exportToZip()).rejects.toThrow(`Agent ${testAgentId} not found`);
      await exportService.cleanup();
    });
  });

  describe('Import Service', () => {
    it('should validate ZIP structure before import', async () => {
      const mockServer = createMockServer();

      // Create invalid ZIP (no manifest)
      const invalidZip = new AdmZip();
      invalidZip.addFile('test.txt', Buffer.from('test'));

      const importService = new AgentImportService(testAgentId, mockServer);

      await expect(importService.importFromZip(invalidZip.toBuffer())).rejects.toThrow(
        'Invalid export: manifest.json not found'
      );

      await importService.cleanup();
    });

    it('should validate manifest version', async () => {
      const mockServer = createMockServer();

      // Create ZIP with invalid version
      const zip = new AdmZip();
      const manifest = {
        version: '99.0.0',
        agentId: 'source-agent',
        agentName: 'Test',
        tables: ['agent'],
        fileCount: 0,
      };

      zip.addFile('manifest.json', Buffer.from(JSON.stringify(manifest)));
      zip.addFile('database/agent.json', Buffer.from('[]'));

      const importService = new AgentImportService(testAgentId, mockServer);

      await expect(importService.importFromZip(zip.toBuffer())).rejects.toThrow(
        'Unsupported export version: 99.0.0'
      );

      await importService.cleanup();
    });

    it('should successfully import valid ZIP', async () => {
      const mockServer = createMockServer();

      // Create valid ZIP
      const zip = new AdmZip();
      const manifest = {
        version: '1.0.0',
        exportedAt: new Date().toISOString(),
        agentId: 'source-agent',
        agentName: 'Test Agent',
        tables: [
          'agent',
          'entities',
          'worlds',
          'rooms',
          'participants',
          'memories',
          'relationships',
          'tasks',
          'server_agents',
        ],
        fileCount: 0,
        elizaVersion: '1.0.0',
      };

      zip.addFile('manifest.json', Buffer.from(JSON.stringify(manifest)));

      // Add all required database files
      const agentData = [
        {
          id: 'source-agent',
          name: 'Test Agent',
          enabled: true,
          createdAt: new Date().toISOString(),
          updatedAt: new Date().toISOString(),
        },
      ];

      zip.addFile('database/agent.json', Buffer.from(JSON.stringify(agentData)));

      // Add empty arrays for other tables
      const emptyTables = [
        'entities',
        'worlds',
        'rooms',
        'participants',
        'memories',
        'relationships',
        'tasks',
        'server_agents',
      ];
      for (const table of emptyTables) {
        zip.addFile(`database/${table}.json`, Buffer.from('[]'));
      }

      const importService = new AgentImportService(testAgentId, mockServer);

      const result = await importService.importFromZip(zip.toBuffer(), {
        overwrite: true,
        validateManifest: true,
      });

      expect(result.success).toBe(true);
      expect(result.agentId).toBe(testAgentId);
      expect(result.agentName).toBe('Test Agent');
      expect(result.tablesImported.length).toBe(9);

      await importService.cleanup();
    });
  });

  describe('Export/Import Cycle', () => {
    it('should maintain data integrity through export/import cycle', async () => {
      // Step 1: Create export data
      const exportZip = new AdmZip();

      const manifest = {
        version: '1.0.0',
        exportedAt: new Date().toISOString(),
        agentId: 'cycle-test-agent',
        agentName: 'Cycle Test Agent',
        tables: [
          'agent',
          'entities',
          'worlds',
          'rooms',
          'participants',
          'memories',
          'relationships',
          'tasks',
          'server_agents',
        ],
        fileCount: 1,
        elizaVersion: '1.0.0',
      };

      exportZip.addFile('manifest.json', Buffer.from(JSON.stringify(manifest, null, 2)));

      // Add agent with complex data
      const agentData = [
        {
          id: 'cycle-test-agent',
          name: 'Cycle Test Agent',
          enabled: true,
          bio: ['Test bio line 1', 'Test bio line 2'],
          settings: { theme: 'dark', language: 'en' },
          metadata: { version: 1, tags: ['test', 'cycle'] },
          createdAt: new Date().toISOString(),
          updatedAt: new Date().toISOString(),
        },
      ];

      exportZip.addFile('database/agent.json', Buffer.from(JSON.stringify(agentData, null, 2)));

      // Add memories with embeddings
      const memories = [
        {
          id: 'memory-1',
          agentId: 'cycle-test-agent',
          content: { text: 'Test memory with embedding' },
          embedding: new Array(768).fill(0.1),
          metadata: { type: 'test' },
          createdAt: Date.now(),
        },
      ];

      exportZip.addFile('database/memories.json', Buffer.from(JSON.stringify(memories, null, 2)));

      // Add other required tables
      const emptyTables = [
        'entities',
        'worlds',
        'rooms',
        'participants',
        'relationships',
        'tasks',
        'server_agents',
      ];
      for (const table of emptyTables) {
        exportZip.addFile(`database/${table}.json`, Buffer.from('[]'));
      }

      // Add a test file
      exportZip.addFile('uploads/test.txt', Buffer.from('test file content'));

      const zipBuffer = exportZip.toBuffer();

      // Step 2: Verify the export
      const verifyZip = new AdmZip(zipBuffer);
      const entries = verifyZip.getEntries();

      expect(entries.length).toBe(11); // manifest + 9 tables + 1 upload

      // Verify manifest
      const manifestEntry = verifyZip.getEntry('manifest.json');
      const readManifest = JSON.parse(verifyZip.readAsText(manifestEntry!));
      expect(readManifest.agentId).toBe('cycle-test-agent');

      // Verify agent data
      const agentEntry = verifyZip.getEntry('database/agent.json');
      const readAgent = JSON.parse(verifyZip.readAsText(agentEntry!));
      expect(readAgent[0].bio).toEqual(['Test bio line 1', 'Test bio line 2']);
      expect(readAgent[0].settings.theme).toBe('dark');

      // Verify memories
      const memoriesEntry = verifyZip.getEntry('database/memories.json');
      const readMemories = JSON.parse(verifyZip.readAsText(memoriesEntry!));
      expect(readMemories[0].embedding.length).toBe(768);
      expect(readMemories[0].content.text).toBe('Test memory with embedding');

      // Verify file
      const fileEntry = verifyZip.getEntry('uploads/test.txt');
      expect(fileEntry?.getData().toString()).toBe('test file content');
    });
  });
});
